### ISSUE TITLE
Emoji support in post

### ISSUE THREAD

**Issue by Mazzibig** [SimpleMachines/SMF#7131]
#### Description
If post text contains emoji with Variation Selector-16 (for example ❗️ U+2757U+FE0F) it has error ''The message body was left empty.'' on posting or preview
In Admin-Error Log error
`2: mb_ereg_replace_callback(): mbregex compile err: invalid character property name {Regional_Indicator}`
and error stack
`#1: mb_ereg_replace_callback()
Call from /..../Sources/Subs-Charset.php in line 3297`
### Steps to reproduce
1. Copy emoji ❗️
2. Paste in post body
3. Click "Preview"

### Environment (complete as necessary)
- Version/Git revision: 2.1RC4 Git https://github.com/SimpleMachines/SMF2.1/commit/aa18d874d3778566c47db02ea5f2ae3f05b10aa1

**Comment by albertlast** [comment]
can't replicate on pg,
maybe a mysql none mb4 issue.
could be fixed by the mb4 pr for mysql.

**Comment by Sesquipedalian** [comment]
What PHP version does this happen on?

> can't replicate on pg,
> 
> maybe a mysql none mb4 issue.
> 
> could be fixed by the mb4 pr for mysql.

No, this is happening because something is missing from PHP's mb_string extension on @Mazzibig's server.

**Comment by sbulen** [comment]
I cannot replicate this on current GH code.

As Sesq suggests, this is likely a php build issue.  It could also possibly be an issue with a failed utf8 conversion somewhere, so settings/db are wrong.

Was this forum upgraded from 2.0 to 2.1RC4? Or installed?

**Comment by Oldiesmann** [comment]
Works fine on my server on a 2.1 RC4 forum

**Comment by jdarwood007** [comment]
Works on my test site.  I'm guessing an outdated pcre library.

**Comment by Mazzibig** [comment]
> What PHP version does this happen on?

7.2.32 mbstring enabled
<html><body>


mbstring |  
-- | --
Multibyte Support | enabled
Multibyte string engine | libmbfl
HTTP input encoding translation | disabled
libmbfl version | 1.3.2
oniguruma version | 6.3.0
Multibyte (japanese) regex support | enabled
Multibyte regex (oniguruma) backtrack check | On
Multibyte regex (oniguruma) version | 6.3.0


</body>
</html>

<html><body>


pcre |  
-- | --
PCRE (Perl Compatible Regular Expressions) Support | enabled
PCRE Library Version | 8.39 2016-06-14
PCRE JIT Support | enabled


</body>
</html>

Emoji without Variation Selector-16 works fine. But with Variation Selector-16 rise error.
In other words,
❗ (U+2757) - working
❗️ (U+2757U+FE0F) - isn`t working

**Comment by Sesquipedalian** [comment]
> oniguruma version	6.3.0

That appears to be the problem. That outdated version of the Oniguruma library does not support some of the character classes used in this regular expression:
https://github.com/SimpleMachines/SMF2.1/blob/a29e75b4e8b31d65546eb39a79d88ffbc85a6374/Sources/Subs-Charset.php#L3265-L3285

A dirty workaround would be to add a test to [this line](https://github.com/SimpleMachines/SMF2.1/blob/a29e75b4e8b31d65546eb39a79d88ffbc85a6374/Sources/Subs-Charset.php#L3246) to check whether the version of the Oniguruma library is at least 6.9.0, so that we can bail out if not. For PHP 7.4+, that'd just be a matter of checking the value of the MB_ONIGURUMA_VERSION constant, but that constant doesn't exist in earlier versions of PHP, so we'd probably have to do something silly like parsing the output of phpinfo().

A better solution would be to replace the character classes that are not supported by old versions of Oniguruma (e.g. `\p{Regional_Indicator}`, `\p{Emoji}`, etc.) with manually specified ones  (e.g. `[\x{1F1E6}-\x{1F1FF}]` in place of `\p{Regional_Indicator}`). However, since classes like `\p{Emoji}` grow with every new version of Unicode, we would definitely want an automated system to build those for us. Fortunately, I added an [internal development file](https://github.com/SimpleMachines/SMF2.1/pull/7124/files#diff-13b0c5619a3de054f18d4097f183af2893d53cd87f812ea07dc6a4ec31cd593e) in #7124 to do exactly that sort of task, so once that PR has been merged it should be quite straightforward to add a bit more code to automatically build these regular expression classes for us.

**Comment by jdarwood007** [comment]
Well we don't need it to do a full support without the proper library correct?  Just have it do a min support?

FYI, Mine on my server is: 6.9.4 (PHP 8.0)
I have another server with 6.8.2 (PHP 7.4)

I actually have no idea how to change/build different versions.  @Mazzibig Is this from a shared hosting environment?

**Comment by jdarwood007** [comment]
Ahh found it:
```
$ yum list installed | grep oniguruma
oniguruma.x86_64                   6.8.2-1.el7                         
oniguruma-devel.x86_64             6.8.2-1.el7                         
```
```
$ apt list --installed | grep libonig
libonig5/focal,now 6.9.4-1 amd64 [installed,automatic]
```
So that explains them..

**Comment by jdarwood007** [comment]
@LiroyvH @VBGAMER45,

You guys run shared servers.  What is the versions for oniguruma showing in PHP info on some of your servers?

Mostly interested in knowing what the oniguruma looks like to understand what sort of stuff we may expect to see in public and if we should add it to our min requirements and/or combat it.

**Comment by LiroyvH** [comment]
@jdarwood007 
> What is the versions for oniguruma showing in PHP info on some of your servers?

6.9.7 on PHP 7.4. Will check other PHP versions (lower and higher) later and let ya know.

**Comment by VBGAMER45** [comment]
I am on dedicated  servers windows plus apache running php 7.4
PHP Version 7.4.24
Multibyte regex (oniguruma) version 	6.9.5

**Comment by VBGAMER45** [comment]
@jdarwood007 

cpanel linux servers I have the following:

PHP Version 7.4.24
Multibyte regex (oniguruma) version 	6.9.7 

PHP Version 8.0.11
Multibyte regex (oniguruma) version 	6.9.7 


PHP Version 7.3.31
Multibyte regex (oniguruma) version 	6.9.4 

PHP Version 7.2.34
Multibyte regex (oniguruma) version 	6.3.0 

PHP Version 7.1.33
Multibyte regex (oniguruma) version 	5.9.6 

PHP Version 7.0.33
Multibyte regex (oniguruma) version 	5.9.6

**Comment by Mazzibig** [comment]
We upgrade our environment to PHP 7.4 to fix this bug. Thanks for support!

**Comment by Sesquipedalian** [comment]
Thanks for the further info, everyone. Based on this, it appears that we really can't assume anything about the installed Oniguruma version.

> Well we don't need it to do a full support without the proper library correct?  Just have it do a min support?

This is true, but it really isn't very difficult to replace shorthand character classes like `\p{Regional_Indicator}` with normal character classes like `[\x{1F1E6}-\x{1F1FF}]`, which will work with any version of Oniguruma (or PCRE, for that matter). Indeed, a one-time fix would be trivial to implement. 

The only theoretically difficult part would be keeping those character classes up to date when new versions of Unicode are released. But as I mentioned earlier, a few additions to the other/update_unicode_data.php file that I added in #7124 can take care of that. Whenever the Unicode Consortium releases a new version, we just run `php -f other/update_unicode_data.php` on the command line, commit the changes, and we're good to go.

**Comment by jdarwood007** [comment]
Well just committing the changes is one thing, after final they have to roll through package manager updates as well.

https://www.php.net/manual/en/migration74.constants.php
According to this, in 7.4 they added MB_ONIGURUMA_VERSION.  So we could for anything with the constant missing (thus below 7.4) or below the 9.6.0 as you mentioned to fall back to the hard coded setup `[\x{1F1E6}-\x{1F1FF}]`, whereas anything above that we can use the `\p{Regional_Indicator}`.

We should have our recommend requirements say the recommended version for this package.

**Comment by sbulen** [comment]
We don't actually need to know all these character classes, correct?  

We only need to know which characters are non-printable...  We only need to identify those, correct?

This is why I don't like rewriting code like this on a perpetual ongoing basis this close to final.

Good ideas, of course.  But good ideas destabilize code.  There is a time for good ideas, and there is a time for stable code.

**Comment by Sesquipedalian** [comment]
> Well just committing the changes is one thing, after final they have to roll through package manager updates as well.

Not a problem, I think. If you look at #7124, you'll see that I moved all the data out of Subs-Charset.php and into separate files in Sources/Unicode/*.php. Because those data files are never, ever meant to be modified by anyone else, we will just need a simple `<require-dir name="Unicode" destination="$sourcedir" />` in package_info.xml.

> So we could for anything with the constant missing (thus below 7.4) or below the 9.6.0 as you mentioned to fall back to the hard coded setup `[\x{1F1E6}-\x{1F1FF}]`, whereas anything above that we can use the `\p{Regional_Indicator}`.

If we're creating the full character classes anyway, there's no point using the shorthand ones. Once we have the full character classes, we don't need to care about library versions at all.

The good news is that, as I said, it wasn't very hard to extend the other/update_unicode_data.php file I added in #7124 so that it builds these regular expressions for us. I've already finished writing the code. I just need to test it a bit more, and then once #7124 has been merged, I'll submit the PR.

**Comment by Sesquipedalian** [comment]
> We don't actually need to know all these character classes, correct?
> 
> We only need to know which characters are non-printable... We only need to identify those, correct?

In a world where there were no non-European scripts and no emoji characters, that would be (almost) true. But in the world of Unicode, things are more complicated. 🙁 

Basically, the problem is that there are several classes of characters in Unicode for which all of the following statements are true:

1. They are invisible in and of themselves.
2. They cause visible, semantically significant differences in surrounding text *in certain contexts*.
3. In other contexts, they make no visible difference.
4. Any of them can be inserted at any arbitrary location in any string of text.

These character classes include formatting characters, variation selectors, emoji modifiers, etc., etc., etc.

Because of statement 2, we can't just remove all occurrences of these invisible characters or else bad things will happen. But because of statements 3 and 4, these characters can easily cause all sorts of problems when misused or abused. Therefore, dealing with these characters correctly requires paying attention to their context rather than a one-size-fits-all approach. It's a pain the butt, but at least the Unicode documentation provides guidance for this, which I followed in the design of the utf8_sanitize_invisibles() function.
